feat: add new REST API endpoints for fetching resource content + metadata

jfriedenstab opened the pull request:
## Description
Add new REST API endpoints for fetching the content and metadata of a deployed RPA resource.

## Related issues

closes #25281

korthout commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1869,6 +1869,78 @@ paths:
+      summary: Get resource
+      description: |
+        Returns a deployed resource.
+        :::note
+        Currently, this endpoint only supports RPA resources.
+        :::
```
👍 It took me some time to verify whether these note admonitions are correctly rendered by our docs, because they aren't used in the current published docs, but they work!

🔧 I think we can increase the visibility to [INFO or WARNING](https://docusaurus.io/docs/markdown-features/admonitions).

jfriedenstab replied:
Actually, they are already used in this `rest-api.yaml` file for a couple of endpoints that are marked as `alpha` 🙂 (e.g. https://docs.camunda.io/docs/8.7/apis-tools/camunda-api-rest/specifications/query-decision-definitions-alpha/).

I've increased the visibility to `INFO` (I wouldn't say that it's worth a `WARNING`).

korthout commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1869,6 +1869,78 @@ paths:
+        - name: resourceKey
+          in: path
+          required: true
+          description: The assigned key of the resource, which acts as a unique identifier for this resource.
```
🔧 All keys in Camunda 8 are assigned. Let's align this to the other usage of the term key in resources: 

```suggestion
          description: The unique key identifying the resource.
```

jfriedenstab replied:
I took the description from the `processDefinitionKey` in [this endpoint](https://docs.camunda.io/docs/next/apis-tools/camunda-api-rest/specifications/get-process-definition/), but I'm happy to use a shorter one.

korthout commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -4915,6 +4987,28 @@ components:
+      type: object
+      properties:
+        resourceName:
+          description: Resource name of this resource.
```
🔧 Let's align to other endpoints

```suggestion
          description: The resource name from which this resource was parsed.
```

korthout commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -4915,6 +4987,28 @@ components:
+          description: Resource name of this resource.
+          type: string
+        version:
+          description: Version of this resource.
```
🔧 Let's align to other endpoints:

```suggestion
          description: The assigned resource version.
```

korthout commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -4915,6 +4987,28 @@ components:
+          type: integer
+          format: int32
+        versionTag:
+          description: Version tag of this resource.
```
🔧 Let's align to other endpoints:

```suggestion
          description: The version tag of the resource, as parsed during deployment.
```

jfriedenstab replied:
I noticed that we use "as parsed during deployment" in some other places, but I'm not sure if it's really useful here, given that the version tag is a custom string provided by the user at development time.

korthout replied:
I don't have a strong opinion on this. If you look around the rest-api doc, the usage of the term ID is anyways not documented consistently. However, I'd rather see it aligned on any of the currently used conventions rather than adding another.

korthout commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/ResourceController.java:
```diff
@@ -58,6 +60,18 @@ public CompletableFuture<ResponseEntity<Object>> deleteResource(
         .fold(RestErrorMapper::mapProblemToCompletedResponse, this::delete);
   }
 
+  @CamundaGetMapping(path = "/resources/{resourceKey}")
+  public CompletableFuture<ResponseEntity<Object>> getResource(
+      @PathVariable final long resourceKey) {
+    return getResource(new ResourceFetchRequest(resourceKey));
+  }
+
+  @CamundaGetMapping(path = "/resources/{resourceKey}/content")
+  public CompletableFuture<ResponseEntity<Object>> getResourceContent(
+      @PathVariable final long resourceKey) {
+    return getResourceContent(new ResourceFetchRequest(resourceKey));
+  }
```
🔧 The methods called by these methods look a lot alike. There's some re-use possible:

```suggestion
  @CamundaGetMapping(path = "/resources/{resourceKey}")
  public CompletableFuture<ResponseEntity<Object>> getResource(
      @PathVariable final long resourceKey) {
    return RequestMapper.executeServiceMethod(
        () -> fetchResource(resourceKey), ResponseMapper::toGetResourceResponse);
  }

  @CamundaGetMapping(path = "/resources/{resourceKey}/content")
  public CompletableFuture<ResponseEntity<Object>> getResourceContent(
      @PathVariable final long resourceKey) {
    return RequestMapper.executeServiceMethod(
        () -> fetchResource(resourceKey), ResponseMapper::toGetResourceContentResponse);
  }

  private CompletableFuture<ResourceRecord> fetchResource(final long resourceKey) {
    return resourceServices
        .withAuthentication(RequestMapper.getAuthentication())
        .fetchResource(new ResourceFetchRequest(resourceKey));
  }
```

jfriedenstab replied:
Yep, good idea 👍🏻

korthout commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1869,6 +1869,78 @@ paths:
+            application/problem+json:
+              schema:
+                $ref: "#/components/schemas/ProblemDetail"
+        "500":
+          description: An internal error occurred while processing the request.
+          content:
+            application/problem+json:
+              schema:
+                $ref: "#/components/schemas/ProblemDetail"
```
🔧 There's no test case for the 500 responses in both endpoints. However, I think such tests are missing in general (I only found one for decision definition query controller). Might be nice to add, but not mandatory apparently.

jfriedenstab replied:
I added tests for this case.

korthout reviewed (approved):
Nicely done @jfriedenstab 🚀 

👍 I found no flaws in the implementation
👍 Both happy and failure cases are covered by tests

🔧 I have a few suggestions here and there. Please have a look, perhaps they are helpful.

LGTM 👍

jfriedenstab commented:
Thanks a lot for your review, @korthout! I've addressed your comments.

By the way, I needed to add https://github.com/camunda/camunda/pull/28223/commits/4e93af842bf4c3b2b595e36a9c24df5d682546af which I missed in https://github.com/camunda/camunda/pull/27981.

korthout reviewed (approved):
Thanks for the changes @jfriedenstab 👍 

LGTM 💯

backport-action commented:
Created backport PR for `main`:
- #28674 with remaining conflicts!

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin backport-28223-to-main
git worktree add --checkout .worktree/backport-28223-to-main backport-28223-to-main
cd .worktree/backport-28223-to-main
git reset --hard HEAD^
git cherry-pick -x d7e6c6b76642d9fd493fa942f38461bf14d429a3 e402a846269a0771699552c01e7ec78ac4bb5a69
git push --force-with-lease
```
